docs/fsharp/whats-new/fsharp-9.md

```diff
@@ -0,0 +1,514 @@
+let len (str: string | null) =
+    let s = nullArgCheck "str" str // Returns a non-null string
+    s.Length  // binds a non-null result
+
+// Declared type at let-binding
+let maybeAValue: string | null = hopefullyGetAString()
+
+// Array type signature
+let f (arr: (string | null)[]) = ()
+
+// Generic code, note 'T must be constrained to be a reference type
+let findOrNull (index: int) (list: 'T list) : 'T | null when 'T : not struct =
+    match List.tryItem index list with
+    | Some item -> item
+    | None -> null
+```
+
+## Discriminated union `.Is*` properties
+
+Discriminated unions now have auto-generated properties for each case, allowing you to check if a value is of a particular case. For example, for the following type:
+
+```fsharp
+type Contact =
+    | Email of address: string
+    | Phone of countryCode: int * number: string
+
+type Person = { name: string; contact: Contact }
+```
+
+Previously, you had to write something like:
+
+```fsharp
+let canSendEmailTo person =
+    match person.contact with
+    | Email _ -> true
+    | _ -> false
+```
+
+Now, you can instead write:
+
+```fsharp
+let canSendEmailTo person =
+    person.contact.IsEmail
+```
+
+## Partial active patterns can return `bool` instead of `unit option`
+
+Previously, partial active patterns returned `Some ()` to indicate a match and `None` otherwise. Now, they can also return `bool`.
+
+For example, the active pattern for the following:
+
+```fsharp
+match key with
+| CaseInsensitive "foo" -> ...
+| CaseInsensitive "bar" -> ...
+```
+
+Was previously written as:
+
+```fsharp
+let (|CaseInsensitive|_|) (pattern: string) (value: string) =
+    if String.Equals(value, pattern, StringComparison.OrdinalIgnoreCase) then
+        Some ()
+    else
+        None
+```
+
+Now, you can instead write:
+
+```fsharp
+let (|CaseInsensitive|_|) (pattern: string) (value: string) =
+    String.Equals(value, pattern, StringComparison.OrdinalIgnoreCase)
+```
+
+## Prefer extension methods to intrinsic properties when arguments are provided
+
+To align with a pattern seen in some .NET libraries, where extension methods are defined with the same names as intrinsic properties of a type, F# now resolves these extension methods instead of failing the type check.
+
+Example:
+
+```fsharp
+type Foo() =
+    member val X : int = 0 with get, set
+
+[&ltExtension>]
+type FooExt =
+    [&ltExtension>]
+    static member X (f: Foo, i: int) = f.X <- i; f
+
+let f = Foo()
+
+f.X(1) // We can now call the extension method to set the property and chain further calls
+```
+
+## Empty-bodied computation expressions
+
+F# now supports empty [computation expressions](../language-reference/computation-expressions.md).
+
+```fsharp
+let xs = seq { } // Empty sequence
+```
+
+```fsharp
+let html =
+    div {
+        p { "Some content." }
+        p { } // Empty paragraph
+    }
+```
+
+Writing an empty computation expression will result in a call to the CE builder's `Zero` method.
```

Comment:
Should "CE" be spelled out as "computation expression"? (I don't know if that's a common term for F# developers)

Reply:
I think it's a common term, but I think it should be spelled out